File: src-wails/dialog.go
package main

import (
	"strings"

	"github.com/wailsapp/wails/v2/pkg/runtime"
)

// PickDirectory/PickFile/PickFiles back the frontend's plugin-dialog
// open() shim (src/lib/wailsCompat/dialog.ts) with Wails' native dialog
// runtime calls. filterName/filterExts mirror Tauri's `filters: [{name,
// extensions}]` (single filter group — every call-site in src/ uses one).

func dialogFilters(filterName string, filterExts []string) []runtime.FileFilter {
	if filterName == "" || len(filterExts) == 0 {
		return nil
	}
	pattern := "*." + strings.Join(filterExts, ";*.")
	return []runtime.FileFilter{{DisplayName: filterName, Pattern: pattern}}
}

// trimQuotes strips literal surrounding double-quote characters that
// macOS's NSOpenPanel-backed dialog can embed in the returned path.
func trimQuotes(s string) string {
	return strings.Trim(s, "\"")
}

func (a *App) PickDirectory() (string, error) {
	path, err := runtime.OpenDirectoryDialog(a.ctx, runtime.OpenDialogOptions{})
	return trimQuotes(path), err
}

func (a *App) PickFile(filterName string, filterExts []string) (string, error) {
	path, err := runtime.OpenFileDialog(a.ctx, runtime.OpenDialogOptions{
		Filters: dialogFilters(filterName, filterExts),
	})
	return trimQuotes(path), err
}

func (a *App) PickFiles(filterName string, filterExts []string) ([]string, error) {
	paths, err := runtime.OpenMultipleFilesDialog(a.ctx, runtime.OpenDialogOptions{
		Filters: dialogFilters(filterName, filterExts),
	})
	for i, p := range paths {
		paths[i] = trimQuotes(p)
	}
	return paths, err
}

File: src/lib/wailsCompat/dialog.ts
// Shim for "@tauri-apps/plugin-dialog"'s open(), backed by Wails' native
// dialog runtime calls (PickDirectory/PickFile/PickFiles in dialog.go).
import * as App from "../../../burrow-wails/burrow/frontend/wailsjs/go/main/App";

interface OpenOptions {
  directory?: boolean;
  multiple?: boolean;
  filters?: { name: string; extensions: string[] }[];
  defaultPath?: string;
}

export async function open(opts: OpenOptions = {}): Promise<string | string[] | null> {
  try {
    if (opts.directory) {
      const dir = await App.PickDirectory();
      return dir || null;
    }
    const filter = opts.filters?.[0];
    const name = filter?.name ?? "";
    const exts = filter?.extensions ?? [];
    if (opts.multiple) {
      const files = await App.PickFiles(name, exts);
      return files && files.length > 0 ? files : null;
    }
    const file = await App.PickFile(name, exts);
    return file || null;
  } catch (err) {
    console.warn("[wails-compat] dialog open() failed:", err);
    return null;
  }
}
